File: donut.py
import math
import time
import os
import shutil

def render_frame(A, B, screen_width, screen_height):
    b = [' '] * (screen_width * screen_height)
    z = [0] * (screen_width * screen_height)

    size_factor = 1.5 

    x_offset = screen_width // 2
    y_offset = screen_height // 2
    x_scale = int(min(screen_width, screen_height) * size_factor / 2)
    y_scale = int(x_scale * 0.7) 

    for j in range(0, 628, 7):
        for i in range(0, 628, 2):
            c = math.sin(i)
            d = math.cos(j)
            e = math.sin(A)
            f = math.sin(j)
            g = math.cos(A)
            h = d + 2
            D = 1 / (c * h * e + f * g + 5)
            l = math.cos(i)
            m = math.cos(B)
            n = math.sin(B)
            t = c * h * g - f * e
            x = int(x_offset + x_scale * D * (l * h * m - t * n))
            y = int(y_offset + y_scale * D * (l * h * n + t * m))
            o = int(x + screen_width * y)
            N = int(8 * ((f * e - c * d * g) * m - c * d * e - f * g - l * d * n))
            if y < screen_height and y > 0 and x > 0 and x < screen_width and D > z[o]:
                z[o] = D
                b[o] = ".,-~:;=!*#$@"[N if N > 0 else 0]

    return ''.join(b)

def clear_screen():
    os.system('cls' if os.name == 'nt' else 'clear')

def get_terminal_size():
    return shutil.get_terminal_size()

def main():
    A = 0
    B = 0
    try:
        while True:
            clear_screen()
            width, height = get_terminal_size()
            frame = render_frame(A, B, width, height - 1) 
            print('\n'.join(frame[i:i+width] for i in range(0, len(frame), width)))
            A += 0.04
            B += 0.02
            time.sleep(0.03)
    except KeyboardInterrupt:
        print("\nDonut animation stopped.")

if __name__ == "__main__":
    main()

File: donut.c
#include <stdio.h>
#include <math.h>
#include <string.h>
#include <unistd.h>
#include <signal.h>
#include <sys/ioctl.h>

volatile sig_atomic_t keep_running = 1;

void handle_sigint(int sig) {
    keep_running = 0;
}

void render_frame(float A, float B, int screen_width, int screen_height, char* output) {
    float size_factor = 1.5;
    int x_offset = screen_width / 2;
    int y_offset = screen_height / 2;
    int x_scale = (int)(fmin(screen_width, screen_height) * size_factor / 2);
    int y_scale = (int)(x_scale * 0.7);

    memset(output, ' ', screen_width * screen_height);
    float z[screen_width * screen_height];
    memset(z, 0, sizeof(float) * screen_width * screen_height);

    for (int j = 0; j < 628; j += 7) {
        for (int i = 0; i < 628; i += 2) {
            float c = sin(i * 0.01);
            float d = cos(j * 0.01);
            float e = sin(A);
            float f = sin(j * 0.01);
            float g = cos(A);
            float h = d + 2;
            float D = 1 / (c * h * e + f * g + 5);
            float l = cos(i * 0.01);
            float m = cos(B);
            float n = sin(B);
            float t = c * h * g - f * e;
            int x = (int)(x_offset + x_scale * D * (l * h * m - t * n));
            int y = (int)(y_offset + y_scale * D * (l * h * n + t * m));
            int o = x + screen_width * y;
            int N = (int)(8 * ((f * e - c * d * g) * m - c * d * e - f * g - l * d * n));
            if (y < screen_height && y > 0 && x > 0 && x < screen_width && D > z[o]) {
                z[o] = D;
                output[o] = ".,-~:;=!*#$@"[N > 0 ? N : 0];
            }
        }
    }
}

int main() {
    struct winsize w;
    float A = 0, B = 0;

    signal(SIGINT, handle_sigint);

    while (keep_running) {
        ioctl(STDOUT_FILENO, TIOCGWINSZ, &w);
        int width = w.ws_col;
        int height = w.ws_row - 1;

        char *frame = malloc(width * height);
        render_frame(A, B, width, height, frame);

        printf("\033[2J\033[H");  // Clear screen and move cursor to top-left
        for (int i = 0; i < height; i++) {
            fwrite(frame + i * width, 1, width, stdout);
            fputc('\n', stdout);
        }
        fflush(stdout);

        free(frame);

        A += 0.04;
        B += 0.02;
        usleep(30000);  // Sleep for 30ms
    }

    printf("\nDonut animation stopped.\n");
    return 0;
}
